pkg/auth/handler/ssohandler/auth_handler.go

```diff
@@ -0,0 +1,398 @@
+}
+
+func (f AuthHandlerFactory) ProvideAuthzPolicy() authz.Policy {
+	return policy.AllOf()
+}
+
+// AuthRequestPayload login handler request payload
+type AuthRequestPayload struct {
+	Code         string
+	Scope        sso.Scope
+	EncodedState string
+}
+
+// Validate request payload
+func (p AuthRequestPayload) Validate() error {
+	if p.Code == "" {
+		return skyerr.NewInvalidArgument("Authorization Code is required", []string{"code"})
+	}
+
+	if p.EncodedState == "" {
+		return skyerr.NewInvalidArgument("EncodedState is required", []string{"state"})
+	}
+
+	return nil
+}
+
+// AuthHandler decodes code response and fetch access token from provider.
+//
+// curl http://localhost:3000/sso/<provider>/auth_handler?code=<code>&state=<state>
+//
+// For ux_mode is 'ios' or 'android',
+// it creates a 302 response, and Location points to:
+// myapp://user.skygear.io/sso/{provider}/auth_handler?result=
+//
+// Fox ux_mode is 'web_redirect',
+// it creates a 302 response, and Location points to: sso_callback_url
+// and set cookie in the response.
+//
+// For ux_mode is 'web_popup',
+// it will render a html page and set cookie in the response.
+//
+type AuthHandler struct {
+	TxContext               db.TxContext                `dependency:"TxContext"`
+	AuthContext             coreAuth.ContextGetter      `dependency:"AuthContextGetter"`
+	Provider                sso.Provider                `dependency:"SSOProvider"`
+	OAuthAuthProvider       oauth.Provider              `dependency:"OAuthAuthProvider"`
+	AuthInfoStore           authinfo.Store              `dependency:"AuthInfoStore"`
+	RoleStore               role.Store                  `dependency:"RoleStore"`
+	TokenStore              authtoken.Store             `dependency:"TokenStore"`
+	AuthHandlerHTMLProvider sso.AuthHandlerHTMLProvider `dependency:"AuthHandlerHTMLProvider"`
+	ProviderName            string
+}
+
+func (h AuthHandler) WithTx() bool {
+	return true
+}
+
+func (h AuthHandler) DecodeRequest(request *http.Request) (handler.RequestPayload, error) {
+	payload := AuthRequestPayload{}
+	q := request.URL.Query()
+	payload.Code = q.Get("code")
+	payload.Scope = strings.Split(q.Get("scope"), " ")
+	payload.EncodedState = q.Get("state")
+
+	return payload, nil
+}
+
+func (h AuthHandler) Handler() http.Handler {
+	// reference from APIHandlerToHandler
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		var respErr skyerr.Error
+
+		if h.Provider == nil {
+			respErr = skyerr.NewInvalidArgument("Provider is not supported", []string{h.ProviderName})
+			return
+		}
+
+		defer func() {
+			if respErr != nil {
+				response := handler.APIResponse{
+					Err: respErr,
+				}
+				handler.WriteResponse(rw, response)
+			}
+		}()
+
+		payload, err := h.DecodeRequest(r)
+		if err != nil {
+			respErr = skyerr.MakeError(err)
+			return
+		}
+
+		if err := payload.Validate(); err != nil {
+			respErr = skyerr.MakeError(err)
+			return
+		}
+
+		if h.WithTx() {
```

Comment:
We can open the transaction here directly and remove `WithTx`. This is for `APIHandler` only so we can have a generic setup to decide whether the handler need auto transaction. If it is always true here, we can just remove it, to make the handler simple.